### PR TITLE
Develop upstream sync 230912

### PR DESCRIPTION
**Pull request by draganmladjenovic**

### CODE REVIEW

File: third_party/xla/xla/pjrt/gpu/BUILD [hunk 1]
```diff
@@ -211,7 +212,7 @@ cc_library(
 xla_cc_test(
     name = "se_gpu_pjrt_compiler_test",
     srcs = if_gpu_is_configured(["se_gpu_pjrt_compiler_test.cc"]),
-    tags = tf_cuda_tests_tags(),
+    tags = tf_cuda_tests_tags() + ["no_rocm"],
```
Comment (i-chaochen) [inline third_party/xla/xla/pjrt/gpu/BUILD]: may I ask why `se_gpu_pjrt_compiler_test` is disable? do we have a ticket to track this?

Comment (i-chaochen) [inline third_party/xla/xla/pjrt/gpu/BUILD]: may I ask why `se_gpu_pjrt_compiler_test` is disable? it's PASSED on XLA upstream side, TF should be ok to enable it.

File: third_party/xla/xla/pjrt/gpu/BUILD [hunk 2]
```diff
@@ -96,6 +96,7 @@ xla_cc_test(
         "noasan",
         "nomsan",
         "requires-gpu-nvidia:2",
+        "no_rocm",
```
Comment (i-chaochen) [inline third_party/xla/xla/pjrt/gpu/BUILD]: https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/5736

File: third_party/xla/xla/service/gpu/gemm_rewriter.h [hunk 3]
```diff
@@ -17,18 +17,11 @@ limitations under the License.
...
-#include "tensorflow/compiler/xla/service/gpu/gpu_types.h"
-#include "tensorflow/compiler/xla/service/hlo_pass_interface.h"
-#include "tensorflow/compiler/xla/service/gpu/gpu_types.h"
-=======
 #include "xla/hlo/ir/hlo_instructions.h"
 #include "xla/hlo/ir/hlo_module.h"
 #include "xla/service/gpu/gpu_types.h"
 #include "xla/service/hlo_pass_interface.h"
->>>>>>> upstream/master:third_party/xla/xla/service/gpu/gemm_rewriter.h
+#include "xla/service/gpu/gpu_types.h"
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/gemm_rewriter.h]: is this duplicated to line 22?

File: third_party/xla/xla/service/gpu/matmul_utils.h [hunk 4]
```diff
@@ -45,8 +31,9 @@ limitations under the License.
 #include "xla/stream_executor/blas.h"
 #include "xla/types.h"
 #include "xla/xla_data.pb.h"
-
->>>>>>> upstream/master:third_party/xla/xla/service/gpu/matmul_utils.h
+#if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+#include "rocm/rocm_config.h"
+#endif
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/matmul_utils.h]: this part should be align with upstream's

https://github.com/openxla/xla/blob/main/xla/service/gpu/matmul_utils.h#L39-L45

File: third_party/xla/xla/service/gpu/tests/hlo_to_llvm_ir.cc [hunk 5]
```diff
@@ -17,24 +17,15 @@ limitations under the License.
...
-#include "tensorflow/compiler/xla/service/gpu/target_constants.h"
-#include "tensorflow/compiler/xla/status.h"
-=======
 #include "xla/hlo/ir/hlo_module.h"
 #include "xla/service/gpu/compile_module_to_llvm_ir.h"
 #include "xla/service/gpu/gpu_device_info_for_tests.h"
 #include "xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.h"
+#if TENSORFLOW_USE_ROCM
+#include "tsl/platform/rocm_rocdl_path.h"
+#endif
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/tests/hlo_to_llvm_ir.cc]: this part is duplicate to line 34

File: third_party/xla/xla/stream_executor/cuda/cuda_blas_lt.cc [hunk 6]
```diff
@@ -26,34 +26,21 @@ limitations under the License.
...
-#include "tensorflow/compiler/xla/stream_executor/stream.h"
-=======
 #include "xla/status_macros.h"
 #include "xla/stream_executor/blas.h"
+#if GOOGLE_CUDA
 #include "xla/stream_executor/cuda/cuda_blas.h"
 #include "xla/stream_executor/cuda/cuda_blas_utils.h"
+#else
+#include "xla/stream_executor/rocm/rocm_blas.h"
+#endif
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/cuda/cuda_blas_lt.cc]: we should remove these rocm code from cuda_blas_lt.cc https://github.com/openxla/xla/pull/3953 as it doesn't need anymore.

File: third_party/xla/xla/stream_executor/cuda/cuda_blas_lt.h [hunk 7]
```diff
@@ -26,14 +26,13 @@ limitations under the License.
...
 #endif
 
-#include "tensorflow/compiler/xla/stream_executor/blas.h"
+#include "xla/stream_executor/blas.h"
 #if GOOGLE_CUDA
-#include "tensorflow/compiler/xla/stream_executor/cuda/cuda_blas_utils.h"
+#include "xla/stream_executor/cuda/cuda_blas_utils.h"
 #else
 #if TF_HIPBLASLT
 #include "../rocm/hipblaslt_wrapper.h"
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/cuda/cuda_blas_lt.h]: ditto

File: third_party/xla/xla/stream_executor/rocm/rocm_dnn.cc [hunk 8]
```diff
@@ -74,7 +48,10 @@ limitations under the License.
 #include "tsl/platform/hash.h"
 #include "tsl/util/determinism.h"
 #include "tsl/util/env_var.h"
->>>>>>> upstream/master:third_party/xla/xla/stream_executor/rocm/rocm_dnn.cc
+#include "rocm/rocm_config.h"
+
+#include <hip/hip_fp16.h>
+#include <hip/hip_bfloat16.h>
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/rocm/rocm_dnn.cc]: we need to upstream these

File: third_party/xla/xla/tools/BUILD [hunk 9]
```diff
@@ -7,6 +7,15 @@ load(
...
 )
 load("@local_config_cuda//cuda:build_defs.bzl", "if_cuda")
+load(
+    "@local_tsl//tsl/platform/default:cuda_build_defs_bzl",
+    "if_cuda_is_configured",
+)
+load(
+    "@local_config_rocm//rocm:build_defs.bzl",
+    "if_rocm_is_configured",
+)
```
Comment (i-chaochen) [inline third_party/xla/xla/tools/BUILD]: this `if_rocm_is_configured` might not need anymore as `if_cuda_or_rocm` is used in this BUILD.

File: third_party/xla/xla/service/gpu/cublas_lt_matmul_thunk.cc [hunk 10]
```diff
@@ -13,12 +13,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
-<<<<<<< HEAD:tensorflow/compiler/xla/service/gpu/cublas_lt_matmul_thunk.cc
-#include "tensorflow/compiler/xla/service/gpu/cublas_lt_matmul_thunk.h"
-#if GOOGLE_CUDA || TF_HIPBLASLT
-=======
 #include "xla/service/gpu/cublas_lt_matmul_thunk.h"
->>>>>>> upstream/master:third_party/xla/xla/service/gpu/cublas_lt_matmul_thunk.cc
+#if GOOGLE_CUDA || TF_HIPBLASLT
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/cublas_lt_matmul_thunk.cc]: we shouldn't need this anymore? BTW: the bottom #endif needs to be removed as well

File: third_party/xla/xla/service/gpu/cublas_lt_matmul_thunk.h [hunk 11]
```diff
@@ -87,9 +87,5 @@ class CublasLtMatmulThunk : public Thunk {
 }  // namespace gpu
 }  // namespace xla
 
-<<<<<<< HEAD:tensorflow/compiler/xla/service/gpu/cublas_lt_matmul_thunk.h
 #endif // GOOGLE_CUDA || TF_HIPBLASLT
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/cublas_lt_matmul_thunk.h]: ditto (line-23 needs to be removed as well)

File: third_party/xla/xla/stream_executor/rocm/hip_blas_lt.h [hunk 12]
```diff
@@ -34,15 +33,9 @@ limitations under the License.
 #define HIPBLASLT_C_64F HIPBLAS_R_64F
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/rocm/hip_blas_lt.h]: it seems this part is missing in upstream and we need to upstreaming it.

File: third_party/xla/xla/tools/multihost_hlo_runner/BUILD [hunk 13]
```diff
@@ -1,124 +0,0 @@
-load("@local_tsl//tsl/platform:rules_cc.bzl", "cc_library")
```
Comment (i-chaochen) [inline third_party/xla/xla/tools/multihost_hlo_runner/BUILD]: interesting...they have removed this `multihost_hlo_runner `from TF (it's still in XLA part)

File: tensorflow/compiler/tf2xla/kernels/light_outside_compilation.cc [hunk 14]
```diff
@@ -105,6 +83,11 @@ limitations under the License.
 #include "xla/stream_executor/gpu/gpu_stream.h"
 #include "xla/stream_executor/gpu/gpu_types.h"
 #include "tensorflow/core/common_runtime/gpu/gpu_device.h"
+#if GOOGLE_CUDA
+#define PLATFORM "CUDA"
+#else
+#define PLATFORM "ROCM"
+#endif
```
Comment (i-chaochen) [inline tensorflow/compiler/tf2xla/kernels/light_outside_compilation.cc]: this needs to be upstreaming https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/commit/3708b75f83eec8b19386423ef0d54cf730f31662

File: third_party/xla/xla/tools/multihost_hlo_runner/BUILD (hunk 13 above)
Comment (draganmladjenovic) [reply third_party/xla/xla/tools/multihost_hlo_runner/BUILD]: Nope. That is my bad on second merge try. Git merge gets lost here so you end up with BUILD files missing or duplicated.

File: .bazelrc
Comment (i-chaochen) [inline .bazelrc:1]: is this related to https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/2192 ?

Comment (draganmladjenovic) [reply .bazelrc:1]: Yes. Or sufficiently newer gcc would work also. And the unused-variable part comes from the change tha we should revert anyway and only affects build_rocm_python3. The gpu-gcc.bazelrc ignores this warrning.

**Comment by draganmladjenovic** [comment]
Please retest Ubuntu-CPU

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-CPU please

**Comment by draganmladjenovic** [comment]
Retest Ubuntu-CPU please

File: third_party/xla/xla/service/gpu/gemm_rewriter.cc [hunk 15]
```diff
@@ -1750,8 +1750,6 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
             {ComputationType::kF32, DataType::kComplexFloat, PrimitiveType::C64,
              PrimitiveType::C64, DataType::kComplexFloat},
 
-            // The next 4 may be supported by hipblaslt, but they are not
-            // covered by any unit tests
             {ComputationType::kF16AsF32, DataType::kFloat, PrimitiveType::F32,
```
Comment (i-chaochen) [inline third_party/xla/xla/service/gpu/gemm_rewriter.cc:1753]: is it already coverted by unit test?

File: tensorflow/tools/ci_build/linux/rocm/run_cpu.sh [hunk 16]
```diff
@@ -55,6 +55,6 @@ bazel test \
       -//tensorflow/core/tpu/... \
       -//tensorflow/lite/... \
       -//tensorflow/tools/toolchains/... \
-      -//tensorflow/compiler/xla/service/gpu/...
+      # -//tensorflow/compiler/xla/service/gpu/...
```
Comment (i-chaochen) [inline tensorflow/tools/ci_build/linux/rocm/run_cpu.sh:58]: we can remove this line if it doesn't need it anymore.

File: tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc [hunk 17]
```diff
@@ -12,59 +12,8 @@ build:sigbuild_remote_cache_push --remote_cache="https://storage.googleapis.com/
...
-test:rocmclang --test_env=HSA_TOOLS_LIB=libroctracer64.so --test_sharding_strategy=disabled
-test --build_tests_only --keep_going --test_output=errors --verbose_failures=true
-# Local test jobs has to be 4 because parallel_gpu_execute is fragile, I think
-test --test_timeout=920,2400,7200,9600 --local_test_jobs=4 --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute
-# Give only the list of failed tests at the end of the log
-test --test_summary=short
-=======
 # Build options for GPU Linux
 build --config=release_gpu_linux
->>>>>>> upstream/master
```
Comment (i-chaochen) [inline tensorflow/tools/tf_sig_build_dockerfiles/devel.usertools/gpu.bazelrc:17]: Upstream had a big updated the build https://github.com/tensorflow/tensorflow/commit/d66aeda6ccce56c994116cac49378eb049959a38
 
are we safe on these two bazelrc files? @jayfurmanek

**Review by i-chaochen — APPROVED** [review]
LGTM

**Pull request merged**